### ISSUE TITLE
Failure to produce exact input sequence from output logits

### ISSUE THREAD

**Issue by hxiaoyang** [huggingface/transformers#28337]
### System Info

- `transformers` version: 4.35.2
- Platform: Linux-6.1.58+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.1
- Safetensors version: 0.4.1
- Accelerate version: 0.25.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- Tensorflow version (GPU?): 2.15.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.7.5 (cpu)
- Jax version: 0.4.23
- JaxLib version: 0.4.23
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
prompt = "Please provide a code sample that reproduces the"
input_ids = torch.tensor([tokenizer.encode(prompt)])
with torch.no_grad():
    outputs = model(input_ids)
string = tokenizer.decode(torch.argmax(outputs.logits[0], dim=1).tolist(),skip_special_tokens=True)
print(string)
```
Output: `note the detailed snippet of demonstrces the issue`

### Expected behavior

I was expecting `string` to be `Please provide a code sample that reproduces the issue`, supposing that `issue` is the generated token here. What's the right way to produce the exact input sequence from output logits?

**Comment by ArthurZucker** [comment]
You should use the `generate` function. Otherwise I have no idea what model you are using, but this is expected: the model does not necessarily predict the token you give as an input. Assisted decoding relies on this property.  

```python
prompt = "Please provide a code sample that reproduces the"
from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained("gpt2")
input_ids = tokenizer(prompt, return_tensors = "pt")
string = tokenizer.batch_decode(model.generate(**input_ids).tolist(),skip_special_tokens=True)
print(string)
['Please provide a code sample that reproduces the code.\n\nIf you are interested in contributing to']
```
use `max_new_tokens` if needed.

**Comment by hxiaoyang** [comment]
Thank you

**Comment by hxiaoyang** [comment]
A followup question: What would be the best way to reuse previous hidden states in a conversational / recursive LM. I tried the following but it didn't quite work.
```python
class RecursiveLM:
    def __init__(self, model, tokenizer):
        self.model = model
        self.tokenizer = tokenizer
        self.history = []
        self.past_key_values = None
    
    def __call__(self, prompt, max_length):
        self.history += [prompt]
        input_ids = torch.tensor([self.tokenizer.encode("".join(self.history))])
        generated_ids = []
        for _ in range(max_length):
            with torch.no_grad():
                outputs = self.model(input_ids, past_key_values=self.past_key_values)
            self.past_key_values = outputs.past_key_values
            generated_id = torch.argmax(outputs.logits[0], dim=1).tolist()[-1]
            generated_ids.append(generated_id)
            input_ids = torch.tensor([[generated_id]])
        generated_text = self.tokenizer.decode(generated_ids, skip_special_tokens=True)
        self.history.append(" "+generated_text)
        return generated_text
```
For my use case, it's important that the entire conversation does not get reprocessed again. (Only previous hidden state and new context should be used.) Should I just replace `model(input_ids, past_key_values)` with `model.generate(input_ids, past_key_values)`?

Thanks! @ArthurZucker

**Comment by ArthurZucker** [comment]
By default the `generate` function will use the `past_key_values` instead of re-computing everything which is probably what you are looking for 😉 the past key values still take into account the previous context in an efficient manner. See an answer [here](https://discuss.huggingface.co/t/what-is-the-purpose-of-use-cache-in-decoder/958)

**Comment by hxiaoyang** [comment]
Thank you!
